test/activerecord_spanner_mock_server/read_only_transaction_test.rb

```diff
@@ -111,9 +111,67 @@ def test_multiple_snapshots
+    def test_single_max_staleness
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("max_staleness: 5.45").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.max_staleness
+      assert_equal 5, single_use.read_only.max_staleness.seconds
+      assert_equal 450000000, single_use.read_only.max_staleness.nanos
+    end
+
+    def test_single_exact_staleness
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("exact_staleness:100").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.exact_staleness
+      assert_equal 100, single_use.read_only.exact_staleness.seconds
+      assert_equal 0, single_use.read_only.exact_staleness.nanos
+    end
+
+    def test_single_min_read_timestamp
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("min_read_timestamp: 2021-09-07T14:33:30.1123Z").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.min_read_timestamp
+      time = Google::Cloud::Spanner::Convert.time_to_timestamp Time.xmlschema("2021-09-07T14:33:30.1123Z")
+      assert_equal time, single_use.read_only.min_read_timestamp
+    end
+
+    def test_single_read_timestamp
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("read_timestamp: 2021-09-07T14:33:31Z").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.read_timestamp
+      time = Google::Cloud::Spanner::Convert.time_to_timestamp Time.xmlschema("2021-09-07T14:33:31.000Z")
+      assert_equal time, single_use.read_only.read_timestamp
+    end
+
+    def register_singer_find_by_id_result with_hints = false
+      res = MockServerTests::create_random_singers_result(1)
+      sql = "SELECT  `singers`.* FROM `singers` WHERE `singers`.`id` = @p1 LIMIT @p2" if with_hints
+      sql = "SELECT `singers`.* FROM `singers` WHERE `singers`.`id` = @p1 LIMIT @p2" unless  with_hints
```

Comment:
nits: two spaces before with_hints.

Reply:
Yes, that is intentional (see above). The extra space is automatically added by ActiveRecord, and the mock server is not really doing any SQL parsing, so it sees a SQL string with an extra space as a different query than one with just one space.

Reply:
Actually, I didn't mean the spaces inside the string. What I mean is the ruby code `unless  with_hints`. But it's not a big deal.

Reply:
Ah, sorry, I completely missed that. Fixed.